File: core/src/main/java/com/jetbrains/tmp/learning/serialization/SupportedLanguagesConverter.java
package com.jetbrains.tmp.learning.serialization;

import com.jetbrains.tmp.learning.SupportedLanguages;
import com.thoughtworks.xstream.converters.SingleValueConverter;

import static com.jetbrains.tmp.learning.SupportedLanguages.INVALID;

public class SupportedLanguagesConverter implements SingleValueConverter {
    @Override
    public String toString(Object o) {
        return ((SupportedLanguages) o).getTitle();
    }

    @Override
    public Object fromString(String s) {
        SupportedLanguages language = SupportedLanguages.langOfName(s);
        if (language != INVALID) {
            return language;
        }

        return SupportedLanguages.langOfTitle(s);
    }

    @Override
    public boolean canConvert(Class aClass) {
        return aClass == SupportedLanguages.class;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/serialization/SupportedLanguagesConverter.kt
package org.stepik.core.serialization

import com.thoughtworks.xstream.converters.SingleValueConverter
import org.stepik.core.SupportedLanguages
import org.stepik.core.SupportedLanguages.INVALID

class SupportedLanguagesConverter : SingleValueConverter {
    override fun toString(any: Any): String {
        return (any as SupportedLanguages).title
    }

    override fun fromString(str: String): SupportedLanguages {
        val language = SupportedLanguages.langOfName(str)
        if (language != INVALID) {
            return language
        }

        if (str == "Java") {
            return SupportedLanguages.JAVA7
        }

        return SupportedLanguages.langOfTitle(str)
    }

    override fun canConvert(clazz: Class<*>): Boolean {
        return clazz == SupportedLanguages::class.java
    }
}
